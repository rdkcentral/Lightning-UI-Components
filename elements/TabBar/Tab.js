import Surface from '../../elements/Surface';
import Icon from '../../elements/Icon';
import TextBox from '../../elements/TextBox';
import { withExtensions } from '../../mixins';

import * as styles from './Tab.styles';

class Tab extends Surface {
  static _template() {
    return {
      ...super._template(),
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2,
        Text: {
          type: TextBox,
          mountY: 0.5,
          signals: {
            textBoxChanged: '_onTextBoxChanged'
          }
        }
      }
    };
  }

  static get __themeStyle() {
    return styles;
  }

  static get __componentName() {
    return 'Tab';
  }

  static get properties() {
    return ['icon', 'title'];
  }

  static get tags() {
    return [
      ...super.tags,
      'Content',
      {
        name: 'Icon',
        path: 'Content.Icon'
      },
      {
        name: 'Text',
        path: 'Content.Text'
      }
    ];
  }

  _onTextBoxChanged() {
    this._updateContent();
    this._updateTabSize();
  }

  _update() {
    super._update();
    this._updateIcon();
    this._updateText();
    this._updateContent();
    this._updateTabSize();
  }

  _updateIcon() {
    if (this.icon) {
      const iconPatch = {
        src: this.icon,
        w: this.style.iconSize,
        h: this.style.iconSize,
        y: this._Content.h / 2,
        color: this.style.contentColor
      };
      if (!this.title) {
        iconPatch.mountX = 0.5;
        iconPatch.x = this._Content.w / 2;
      }
      if (this._Icon) {
        this._Icon.patch(iconPatch);
      } else {
        this._Content.patch({
          Icon: {
            type: Icon,
            mountY: 0.5,
            ...iconPatch
          }
        });
      }
    } else {
      this._Content.patch({ Icon: undefined });
    }
  }

  _updateText() {
    const textPatch = {
      content: this.title,
      textStyle: this.style.textStyle,
      y: this._Content.h / 2
    };
    if (this.icon) {
      textPatch.mountX = 0;
      textPatch.x = this._iconW + this.style.iconMarginRight;
    } else {
      textPatch.mountX = 0.5;
      textPatch.x = this._Content.w / 2;
    }
    this._Text.patch(textPatch);
  }

  _updateContent() {
    this._Content.patch({
      w:
        this._iconW +
        (this.title ? this.style.iconMarginRight : 0) +
        this._textW,
      h: Math.max(this._iconH, this._Text.h)
    });
  }

  _updateTabSize() {
    if (this.title || this.icon) {
      this.patch({
        w: this._paddingX * 2 + this._Content.w,
        h: this.style.paddingY * 2 + this._Content.h
      });
    } else {
      this.patch({ w: 0, h: 0 });
    }
  }

  get _textW() {
    return this.title ? this._Text.w : 0;
  }

  get _iconW() {
    return this.icon ? this._Icon.w : 0;
  }

  get _iconH() {
    return this.icon ? this._Icon.h : 0;
  }

  get _paddingX() {
    return this.title ? this.style.paddingX : this.style.noTitlePaddingX;
  }
}

export default withExtensions(Tab);
