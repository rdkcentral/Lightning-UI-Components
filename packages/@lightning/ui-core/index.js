export { default as utils } from './src/utils/index.js';
export * from './src/components/index.js';
export * from './src/globals/index.js';
export * from './src/mixins/index.js';
export * from './src/shaders/index.js';
export * from './src/textures/index.js';
