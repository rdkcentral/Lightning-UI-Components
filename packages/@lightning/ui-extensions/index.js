export * from './src/FocusRing';
export * from './src/DropShadow';
