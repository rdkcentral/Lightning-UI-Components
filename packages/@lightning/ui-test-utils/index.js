export * from './src/lightning-test-renderer';
export * from './src/lightning-test-utils';
export * from './src/lng-test-env';
