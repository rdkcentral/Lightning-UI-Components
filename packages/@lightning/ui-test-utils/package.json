{
  "name": "@lightning/ui-test-utils",
  "version": "1.1.0",
  "type": "module",
  "exports": "./index.js",
  "files": [
    "src/*"
  ],
  "peerDependencies": {
    "@lightningjs/core": "^2.7.1"
  },
  "devDependencies": {
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jest-webgl-canvas-mock": "^0.2.3"
  },
  "dependencies": {
    "@lightning/ui-core": "workspace:^",
    "@lightningjs/core": "^2.7.1"
  }
}
