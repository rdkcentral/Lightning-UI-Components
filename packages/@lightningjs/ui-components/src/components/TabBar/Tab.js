/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Surface from '../Surface';
import Icon from '../Icon';
import TextBox from '../TextBox';
import * as styles from './Tab.styles.js';

export default class Tab extends Surface {
  static _template() {
    return {
      ...super._template(),
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2,
        Text: {
          type: TextBox,
          mountY: 0.5,
          signals: {
            textBoxChanged: '_onTextBoxChanged'
          }
        }
      }
    };
  }

  static get __themeStyle() {
    return styles;
  }

  static get __componentName() {
    return 'Tab';
  }

  static get properties() {
    return ['icon', 'title'];
  }

  static get tags() {
    return [
      ...super.tags,
      'Content',
      {
        name: 'Icon',
        path: 'Content.Icon'
      },
      {
        name: 'Text',
        path: 'Content.Text'
      }
    ];
  }

  _onTextBoxChanged() {
    this._updateContent();
    this._updateTabSize();
  }

  _update() {
    super._update();
    this._updateIcon();
    this._updateText();
    this._updateContent();
    this._updateTabSize();
  }

  _updateIcon() {
    if (this.icon) {
      const iconPatch = {
        src: this.icon,
        w: this.style.iconSize,
        h: this.style.iconSize,
        y: this._Content.h / 2,
        color: this.style.iconColor
      };
      if (!this.title) {
        iconPatch.mountX = 0.5;
        iconPatch.x = this._Content.w / 2;
      }
      if (this._Icon) {
        this._Icon.patch(iconPatch);
      } else {
        this._Content.patch({
          Icon: {
            type: Icon,
            mountY: 0.5,
            ...iconPatch
          }
        });
      }
    } else {
      this._Content.patch({ Icon: undefined });
    }
  }

  _updateText() {
    const textPatch = {
      content: this.title,
      style: { textStyle: this.style.textStyle },
      y: this._Content.h / 2
    };
    if (this.icon) {
      textPatch.mountX = 0;
      textPatch.x = this._iconW + this.style.iconMarginRight;
    } else {
      textPatch.mountX = 0.5;
      textPatch.x = this._Content.w / 2;
    }
    this._Text.patch(textPatch);
  }

  _updateContent() {
    this._Content.patch({
      w:
        this._iconW +
        (this.title ? this.style.iconMarginRight : 0) +
        this._textW,
      h: Math.max(this._iconH, this._Text.h)
    });
  }

  _updateTabSize() {
    if (this.title || this.icon) {
      this.patch({
        w: this._paddingX * 2 + this._Content.w,
        h: this.style.paddingY * 2 + this._Content.h
      });
    } else {
      this.patch({ w: 0, h: 0 });
    }
  }

  get _textW() {
    return this.title ? this._Text.w : 0;
  }

  get _iconW() {
    return this.icon ? this._Icon.w : 0;
  }

  get _iconH() {
    return this.icon ? this._Icon.h : 0;
  }

  get _paddingX() {
    return this.title ? this.style.paddingX : this.style.noTitlePaddingX;
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return this._announce || (this._Text && this._Text.announce);
  }
}
