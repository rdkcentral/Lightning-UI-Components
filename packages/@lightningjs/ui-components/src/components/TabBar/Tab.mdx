import { Canvas, Story } from '@storybook/blocks';
import { Tab } from '.';

# Tab

Tab component to be rendered in a [TabBar](/docs/elements-tabbar--basic).

## Source

https://github.com/rdkcentral/Lightning-UI-Components/blob/develop/packages/@lightningjs/ui-components/src/components/TabBar/Tab.js

## Usage

`Tab` is the recommended component to use when defining `tabs` in a [TabBar](/docs/elements-tabbar--basic) component.

```js
import { TabBar, Tab, Row } from '@lightningjs/ui-components';

class Example extends lng.Component {
  static _template() {
    return {
      Tab: {
        Tab: {
          type: Tab,
          icon: iconPath,
          title: 'Tab'
        }
      }
    };
  }
}
```

<Canvas>
  <Story id="layout-tab--tab" />
</Canvas>

### Properties

| name       | type                                                    | required | default   | description                                                                                                     |
| ---------- | ------------------------------------------------------- | -------- | --------- | --------------------------------------------------------------------------------------------------------------- |
| icon       | string                                                  | false    | undefined | path to icon to prepend tab title with                                                                          |
| title      | string                                                  | false    | undefined | text to display on tab                                                                                          |
| tabContent | <DocsLink id="lng.Component">lng.Component[]</DocsLink> | false    | undefined | tabContent accepts a component, a function which returns a component, or a promise that resolves to a component |

### Style Properties

| name            | type             | description                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| --------------- | ---------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| backgroundColor | string           | background color of tab                                                                                                                                                                                                                                                                                                                                                                                                                              |
| contentColor    | string           | color of icon and/or text                                                                                                                                                                                                                                                                                                                                                                                                                            |
| iconMarginRight | number           | space between icon and text                                                                                                                                                                                                                                                                                                                                                                                                                          |
| iconSize        | number           | width and height of icon                                                                                                                                                                                                                                                                                                                                                                                                                             |
| noTitlePaddingX | number           | horizontal padding of tab when only an icon is rendered                                                                                                                                                                                                                                                                                                                                                                                              |
| paddingX        | number           | horizontal padding of tab                                                                                                                                                                                                                                                                                                                                                                                                                            |
| paddingY        | number           | vertical padding of tab                                                                                                                                                                                                                                                                                                                                                                                                                              |
| radius          | number           | border radius of tab                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| textStyle       | string \| object | Style of text to be displayed. This value will take precedent over the `textStyle` style property. Available values: display1, display2, headline1, headline2, headline3, body1, body2, body3, button1, button2, callout1, caption1. Alternatively an object can be passed in containing any [properties that the Lightning text texture supports](https://lightningjs.io/docs/#/lightning-core-reference/RenderEngine/Textures/Text?id=properties). |
