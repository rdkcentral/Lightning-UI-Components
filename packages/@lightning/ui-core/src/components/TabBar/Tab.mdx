import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../../storybook/index.js';
import { Tab } from '.';

# Tab

Tab component to be rendered in a [TabBar](/docs/elements-tabbar--basic).

## Source

https://github.comcast.com/Lightning/lightning-ui/blob/develop/packages/@lightning/ui-core/src/components/TabBar/Tab.js

## Usage

`Tab` is the recommended component to use when defining `tabs` in a [TabBar](/docs/elements-tabbar--basic) component.

```js
import { TabBar, Tab, Row } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Tab: {
        Tab: {
          type: Tab,
          icon: iconPath,
          title: 'Tab'
        }
      }
    };
  }
}
```

<Canvas>
  <Story id="layout-tab--tab" />
</Canvas>

### Properties

| name  | type   | required | default | description                            |
| ----- | ------ | -------- | ------- | -------------------------------------- |
| icon  | string | false    | -       | path to icon to prepend tab title with |
| title | string | false    | -       | text to display on tab                 |

### Style Properties

| name            | type             | description                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| --------------- | ---------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| backgroundColor | string           | background color of tab                                                                                                                                                                                                                                                                                                                                                                                                                              |
| contentColor    | string           | color of icon and/or text                                                                                                                                                                                                                                                                                                                                                                                                                            |
| iconMarginRight | number           | space between icon and text                                                                                                                                                                                                                                                                                                                                                                                                                          |
| iconSize        | number           | width and height of icon                                                                                                                                                                                                                                                                                                                                                                                                                             |
| noTitlePaddingX | number           | horizontal padding of tab when only an icon is rendered                                                                                                                                                                                                                                                                                                                                                                                              |
| paddingX        | number           | horizontal padding of tab                                                                                                                                                                                                                                                                                                                                                                                                                            |
| paddingY        | number           | vertical padding of tab                                                                                                                                                                                                                                                                                                                                                                                                                              |
| radius          | number           | border radius of tab                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| textStyle       | string \| object | Style of text to be displayed. This value will take precedent over the `textStyle` style property. Available values: display1, display2, headline1, headline2, headline3, body1, body2, body3, button1, button2, callout1, caption1. Alternatively an object can be passed in containing any [properties that the Lightning text texture supports](https://lightningjs.io/docs/#/lightning-core-reference/RenderEngine/Textures/Text?id=properties). |
