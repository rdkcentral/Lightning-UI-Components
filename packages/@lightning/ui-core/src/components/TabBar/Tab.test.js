import { makeCreateComponent } from '@lightning/ui-test-utils';
import Tab from './Tab.js';

const iconPath = '../../assets/images/ic_lightning_white_32.png';

const createComponent = makeCreateComponent(Tab);

describe('Tab', () => {
  let component, testRenderer;

  beforeEach(async () => {
    [component, testRenderer] = createComponent(
      {},
      { spyOnMethods: ['_update'] }
    );
    await component.__updateSpyPromise;
  });

  it('renders', async () => {
    [component, testRenderer] = createComponent(
      {
        title: 'Tab',
        icon: iconPath
      },
      {
        spyOnMethods: ['_update']
      }
    );

    await component.__updateSpyPromise;

    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the title string', () => {
    const title = 'Tab 1';
    component.title = title;
    testRenderer.forceAllUpdates();
    expect(component.announce).toBe(title);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    component.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(component.announce).toBe(overrideString);
  });

  it('should render a tab with just an icon', async () => {
    expect(component._Icon).not.toBeDefined();

    component.icon = iconPath;
    await component.__updateSpyPromise;

    expect(component._Icon).toBeDefined();
  });

  it('should render a tab with just text', async () => {
    const title = 'Tab';
    expect(component._Text.content).toBeUndefined();

    component.title = title;
    await component.__updateSpyPromise;

    expect(component._Text.content).toBe(title);
    expect(component._Text.mountX).toBe(0.5);
  });

  it('should render a tab with an icon and text', async () => {
    const title = 'Tab';

    component.icon = iconPath;
    component.title = title;
    await component.__updateSpyPromise;

    expect(component._Icon).toBeDefined();
    expect(component._Text.content).toBe(title);
    expect(component._Text.mountX).toBe(0);
  });

  it('should remove icon and text from a tab when not defined', async () => {
    const title = 'Tab';

    component.icon = iconPath;
    component.title = title;
    await component.__updateSpyPromise;

    expect(component._Icon).toBeDefined();
    expect(component._Text.content).toBe(title);
    expect(component.w).toBeGreaterThan(0);
    expect(component.h).toBeGreaterThan(0);

    component.icon = undefined;
    component.title = undefined;
    await component.__updateSpyPromise;

    expect(component._Icon).toBeUndefined();
    expect(component._Text.content).toBe('');
    expect(component.w).toBe(0);
    expect(component.h).toBe(0);
  });
});
