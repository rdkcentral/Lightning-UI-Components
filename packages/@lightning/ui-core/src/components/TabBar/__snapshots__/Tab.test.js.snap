// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab renders 1`] = `
{
  "Component": {
    "Background": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Background",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Content": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Icon": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": "fff6f6f9",
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 40,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Icon",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Icon",
          ],
          "type": "Icon",
          "visible": true,
          "w": 40,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Text": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Text": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Text",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Text",
              ],
              "texture": {
                "fontFace": "XfinityBrownBold",
                "fontSize": 28,
                "letterSpacing": -0.2,
                "lineHeight": 40,
                "maxLines": undefined,
                "maxLinesSuffix": "...",
                "text": "Tab",
                "textBaseline": "bottom",
                "textColor": 4294375161,
                "textOverflow": "ellipsis",
                "type": "TextTexture",
                "verticalAlign": "middle",
                "wordWrapWidth": undefined,
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 2,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Text",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Text",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 48,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 40,
      "isComponent": undefined,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Content",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Content",
      ],
      "visible": true,
      "w": 48,
      "x": [Function],
      "y": [Function],
      "zIndex": 0,
    },
  },
}
`;
